pkg/utils/host/host.go

```diff
@@ -80,16 +97,130 @@ func init() {
 	IsHostNetNs = isHostNamespace("net")
 }
 
-func workarounds() error {
-	// No memory limit for eBPF maps
-	if err := rlimit.RemoveMemlock(); err != nil {
+func checkKubectlDebug() error {
+	if HostRoot != "/host" || os.Getenv("KUBERNETES_PORT") == "" {
+		return nil
+	}
+
+	c, err := capability.NewPid2(0)
+	if err != nil {
 		return err
 	}
+	err = c.Load()
+	if err != nil {
+		return err
+	}
+	if c.Get(capability.EFFECTIVE, capability.CAP_SYS_ADMIN) {
+		// if we already have CAP_SYS_ADMIN, we don't need a workaround
+		return nil
+	}
+
+	_, err = os.Stat("/host/run/systemd/private")
+	if err != nil {
+		// if the host does not use systemd, we cannot use this workaround
+		return nil
+	}
+
+	// systemdDbus.NewSystemdConnectionContext() hardcodes the path to the
+	// systemd socket to /run/systemd/private. We need to make sure that this
+	// path exists (if the /run:/run mount was set up correctly). If it doesn't
+	// exist, we create the symlink to /host/run/systemd/private.
+	_, err = os.Stat("/run/systemd/private")
+	if errors.Is(err, os.ErrNotExist) {
+		err := os.MkdirAll("/run/systemd", 0o755)
+		if err != nil {
+			return err
+		}
+
+		err = os.Symlink("/host/run/systemd/private", "/run/systemd/private")
+		if err != nil {
+			return fmt.Errorf("linking /run/systemd/private: %w", err)
+		}
+	} else if err != nil {
+		return fmt.Errorf("statting /run/systemd/private: %w", err)
+	}
+
+	conn, err := systemdDbus.NewSystemdConnectionContext(context.TODO())
+	if err != nil {
+		return fmt.Errorf("connecting to systemd: %w", err)
+	}
+	defer conn.Close()
+
+	runID := uuid.New().String()[:8]
+	unitName := fmt.Sprintf("kubectl-debug-ig-%s.service", runID)
+
+	exit := make(chan os.Signal, 1)
+	signal.Notify(exit, syscall.SIGINT, syscall.SIGTERM)
+
+	statusChan := make(chan string, 1)
+	cmd := []string{
+		fmt.Sprintf("/proc/%d/root/usr/bin/ig", os.Getpid()),
+	}
+	cmd = append(cmd, os.Args[1:]...)
+	properties := []systemdDbus.Property{
```

Comment:
[Documentation for "kubectl debug node"](https://kubernetes.io/docs/tasks/debug/debug-cluster/kubectl-node-debug/):

> Although the container runs in the host IPC, Network, and PID namespaces, the pod isn't privileged.

kubectl debug starts a pod with some privileges but not all. It has the `/host` volume, some host namespaces and some capabilities but not CAP_SYS_ADMIN. And it does not have any option like `--privileged` to give the missing capabilities required by Inspektor Gadget.

So as a workaround, `ig` re-executes itself but on the host in a fully privileged systemd unit.

I am unsure if this workaround should be automatic. Maybe it could prompt the user on the terminal unless an environment variable `$IG_AUTO_REEXEC_IN_SD_UNIT` is set:

```
$ kubectl debug node/minikube-docker -ti --image=albantest.azurecr.io/ig:dev \
    -e IG_AUTO_REEXEC_IN_SD_UNIT=1 -- ig trace exec
```

```
$ kubectl debug node/minikube-docker -ti --image=albantest.azurecr.io/ig:dev -- ig trace exec
Creating debugging pod node-debugger-minikube-docker-9t8br with container debugger on node minikube-docker.
If you don't see a command prompt, try pressing enter.

ig is running in a container without CAP_SYS_ADMIN but with access to systemd.
Reexecute ig in a systemd unit? [yes/no] yes

CONTAINER                                PID        PPID       COMM             RET ARGS
k8s_shell01_shell01_default_40bfae0e-51… 3294896    3294876    ls               0   /bin/ls
```

Reply:
> I find it complicated.

I talked about this with colleagues over lunch and the conclusion was that users would not like if `ig` creates a systemd unit without telling the user, so a flag like `--allow-sdunit` would be better.

At the moment, all this code is implemented in a `init()` function, so before CLI flags are parsed. It would need to be moved to 

> In any case, without the workaround ig would not work while run from kubectl debug?

Correct.

> So, I think this workaround is needed.

Correct. But we should do it in a way which is not behind the user's back.

```
$ kubectl debug node/... -- ig trace exec
Error: container without CAP_SYS_ADMIN detected. Use --allow-sdunit to try again in a privileged systemd unit.
$ kubectl debug node/... -- ig trace exec --allow-sdunit
CONTAINER  PID        PPID       COMM             RET ARGS
foo        3294896    3294876    ls               0   /bin/ls
```

What do you think?

Reply:
The workaround should be described in the documentation, explaining what kubectl debug does, what is missing to be able to run `ig`, and what we are doing to have those missing configurations.

In addition, we should say something like: "To make the user aware of what this workaround, we intentionally let the command fail if `--allow-sdunit`.